### ISSUE TITLE
Suggestion: Dropping support for either autotools or cmake

### ISSUE THREAD

**Issue by C4K3** [umurmur/umurmur#111]
Supporting more than one build system is a waste of resources. It leads to incompatabilities depending on what was compiled with, see e.g. #110. Is there anybody that actually needs either autotools or cmake and can't live without the other?

This has been discussed before in #74. The only downside I can think of is that some packaging scripts will have to be rewritten, but the only packager who chimed in had the opinion that supporting only one build system is best.

**Comment by omor1** [comment]
I agree that supporting multiple build systems is prone to breakage and is a waste of developer resources.

It seems that autotools is currently a bit broken, according to #110, and it's a bit more complicated to maintain, so I think that going with cmake is probably the better idea. I have a different (small) project that uses autotools, but that's mostly because it uses some functions from Gnulib, which only really integrates with autotools (because GNU only cares about their own build system).

**Comment by C4K3** [comment]
I only have experience with autotools, and while it works well when it works, it can be an absolute PITA at times as well.

I decided to check a distros that ship umurmur to see what they use:

* Arch: cmake
* Gentoo: autotools
* Opensuse: autotools
* OpenWRT: autotools

**Comment by fmorgner** [comment]
@C4K3 Thanks for evaluating which build systems are used by different distributions. I agree with your and @omor1 in that managing two independent build systems is overkill. Since i come from a C++ background
I really enjoy working with CMake since it feels much more manageable to me. I think however, that we probably will have to drop CMake as it seems (according to your study) that almost none of the distros use CMake to build umurmur.

Since, as you already mentioned, we only had one maintainer chime in the issue sofar, I will try contacting the maintainers of different distro packages to try and find out why they use autotools instead of CMake.

**Comment by omor1** [comment]
@fmorgner more than likely, the packages already used autotools and they never bothered to update to cmake, as the autotools backend still worked fine (they might have never realized there even was a new build system). They'll update if packaging stops working 😄.

It's probably best to go with what causes the developers the least headache—at the end of the day, _we're_ the ones that are constantly building and rebuilding the software. From a distribution's standpoint, I doubt they truly care what the build system is, as long as it works well and isn't something particularly peculiar or esoteric.

**Comment by C4K3** [comment]
I agree with omor1. But I do think it would be best to deprecate it for at least one release before we remove it, to give some warning. Maybe we could break it in some trivial way (rename autogen.sh.) for the next release.

**Comment by omor1** [comment]
@C4K3 maybe add a warning in `configure.ac`? Something like the following:
```
m4_warn([obsolete], [autotools are deprecated, use cmake instead])
```
That should output a warning when running `auto[re]conf`. I'm not sure if it should be in the `obsolete` category though (it seems to refer more to obsolete autoconf constructs), and haven't actually tested the above code.

**Comment by concatime** [comment]
Ok. It seems I'm not the only one getting trouble with autotools and OpenSSL 1.1.
I'm trying to make a [one scirpt run](https://gist.github.com/concatime/19567a1694dfdc52ceb2e0e022ea9a88/e75382c06b95999b8ac9fdfdef66710d68d11d55) with new clean VPS (e.g. digitalocean), but I can't with autotools.
If someone came with a solution, let me know.

**Comment by C4K3** [comment]
Any issues with autotools and OpenSSL 1.1.0 should have been fixed by #109. If you continue to have problems, open a new issue.

**Comment by concatime** [comment]
Ok, I'm just searching for help to finish it. I want to know, does --with-ssl take only openssl as argument, or the path of openssl source?

**Comment by C4K3** [comment]
Yes, it only takes openssl as an argument, autotools should then find openssl. [The wiki](https://github.com/umurmur/umurmur/wiki/Building) has documentation for all the configuration options.

**Comment by concatime** [comment]
Ok, but how do I tell autotools to pick the openssl 1.1 from /opt/openssl and not the default one (which is I assume OpenSSL 1.0)?

**Comment by C4K3** [comment]
You should be able to do that by setting the LDFLAGS environment variable. But please open a new issue so we don't clutter this one with offtopic chat.

**Comment by unterwulf** [comment]
+1 for autotools. Just because they are very conventional for maintainers. BTW, what was the reason to introduce cmake? Were there something that couldn't be implemented with autotools?

**Comment by C4K3** [comment]
I think we've already decided to go with cmake.

I wasn't around for the introduction of cmake, so I can't speak to that.

**Comment by unterwulf** [comment]
I am just making sure if there a possibility to prevent this transition. Have distros already switched to cmake?

Are there cmake experts among active developers? I have only experience with autotools and I am going to contribute some features that I just don't know how to express in cmake. E.g. I've written a manual page that needs to be expanded with actual paths used on the configure step before installation and I only know how to do it with automake.

**Comment by C4K3** [comment]
Anything is possible. I haven't checked if distros have switched.

As you've probably noticed, umurmur is not very active the past couple years, so depends on your definition of active developers.

**Comment by unterwulf** [comment]
As far as I can see the situation is still the same. Arch is the only distro who switched to cmake.
Here are the links I used to check (I don't use any of these distros, so the links is not necessarily reliable):
* Arch: https://www.archlinux.org/packages/community/x86_64/umurmur/
* Gentoo: https://packages.gentoo.org/packages/media-sound/umurmur
* Opensuse: https://build.opensuse.org/package/show/home%3Adarix%3Aplayground/umurmur
* OpenWRT: https://git.openwrt.org/?p=feed/packages.git;a=blob;f=net/umurmur/Makefile

**Comment by C4K3** [comment]
Ok, so nothing's changed.

I personally don't have a preference for either autotools or cmake. If you prefer autotools and want to commit to maintaining autotools for umurmur, I'll support you 100%.

**Comment by uis246** [comment]
OpenWRT uses CMake

**Comment by troxor** [comment]
Personally I favor CMake, perhaps because "i use arch btw", but agree a deprecation warning would be a good idea either way. This is a good candidate for Discussion to see where various distro maintainers are at these days.
